include/pmacc/kernel/warp.hpp

```diff
@@ -53,66 +53,70 @@ namespace pmacc
 
 
 #    if(__CUDA_ARCH__ >= 300 || BOOST_COMP_HIP)
-        /** broadcast data within a warp
-         *
-         * required PTX ISA >=3.0
+
+        /** broadcast data within a warp without using shared memory
          *
+         * @param mask mask to select threads participating in the operation
          * @param data value to broadcast
          * @param srcLaneId lane id of the source thread
-         * @return value send by the source thread
-         *
-         * \{
          */
-        //! broadcast a 32bit integer
-        DINLINE int32_t warpBroadcast(const int32_t data, const int32_t srcLaneId)
-        {
-#        if(__CUDACC_VER_MAJOR__ >= 9)
-            return __shfl_sync(__activemask(), data, srcLaneId);
-#        else
-            return __shfl(data, srcLaneId);
-#        endif
-        }
+        template<typename T_MaskType, typename T_DataType>
+        DINLINE T_DataType warpBroadcast(T_MaskType const mask, T_DataType const data, int const srcLaneId);
 
-        //! Broadcast a 64bit integer by using 2 32bit broadcasts
-        DINLINE int64_cu warpBroadcast(int64_cu data, const int32_t srcLaneId)
-        {
-            int32_t* const pData = reinterpret_cast<int32_t*>(&data);
-            pData[0] = warpBroadcast(pData[0], srcLaneId);
-            pData[1] = warpBroadcast(pData[1], srcLaneId);
-            return data;
-        }
+        /** check if the first type is equal to any other */
+        template<typename T, typename... T_TypeList>
+        inline constexpr bool matchAnyType = (std::is_same_v<T, T_TypeList> || ...);
 
-        //! Broadcast a 32bit unsigned int
-        DINLINE uint32_t warpBroadcast(const uint32_t data, const int32_t srcLaneId)
-        {
-            return static_cast<uint32_t>(warpBroadcast(static_cast<int32_t>(data), srcLaneId));
-        }
 
-        //! Broadcast a 64bit unsigned int
-        DINLINE uint64_cu warpBroadcast(const uint64_cu data, const int32_t srcLaneId)
+        /** fallback for not nativly supported types
+         *
+         * emulates the brotcast by repeating 32bit warp broatcasts
+         */
+        template<typename T_MaskType, typename T_DataType, typename T_Sfinae = void>
+        struct WarpBroatcast
         {
-            return static_cast<uint64_cu>(warpBroadcast(static_cast<int64_cu>(data), srcLaneId));
-        }
+            DINLINE T_DataType operator()(T_MaskType const mask, T_DataType data, int const srcLaneId) const
+            {
+                static_assert(sizeof(T_DataType) >= sizeof(int) && (sizeof(T_DataType) % sizeof(int)) == 0);
+                int* const pData = reinterpret_cast<int*>(&data);
```

Comment:
The previous implementation was more precise about the type here: `sizeof(int)` is not necessarily 32bit. Should we use `int32_t` here as before?

Reply:
I changed it back to `int` because the native interfaces requires `int` and in principle `int32_t` could be anything.
So the new implementation will guarantee that all works. The old most likely worked but  if we are pedantic than there is no guarantee.